File: services/vector_service.py
import numpy as np
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_
from typing import List, Tuple

from db.models import Chunk, Document
from core.config import settings


class VectorService:
    
    def cosine_similarity(self, a: List[float], b: List[float]) -> float:
        """Calculate cosine similarity between two vectors"""
        a_np = np.array(a)
        b_np = np.array(b)
        
        dot_product = np.dot(a_np, b_np)
        norm_a = np.linalg.norm(a_np)
        norm_b = np.linalg.norm(b_np)
        
        if norm_a == 0 or norm_b == 0:
            return 0.0
        
        return dot_product / (norm_a * norm_b)
    
    async def search_similar_chunks(
        self,
        db: AsyncSession,
        query_embedding: List[float],
        user_id: str,
        limit: int = None
    ) -> List[Tuple[Chunk, Document, float]]:
        """Search for similar chunks using cosine similarity"""
        if limit is None:
            limit = settings.max_chunks_context
        
        # Get all chunks from ready documents for the user
        result = await db.execute(
            select(Chunk, Document)
            .join(Document, Chunk.document_id == Document.id)
            .where(
                and_(
                    Document.status == "ready",
                    Document.uploader_id == user_id,
                    Chunk.embedding.isnot(None)
                )
            )
        )
        
        chunks_with_docs = result.all()
        
        # Calculate similarities
        similarities = []
        for chunk, document in chunks_with_docs:
            if chunk.embedding:
                similarity = self.cosine_similarity(query_embedding, chunk.embedding)
                similarities.append((chunk, document, similarity))
        
        # Sort by similarity and return top results
        similarities.sort(key=lambda x: x[2], reverse=True)
        
        # Filter by threshold and limit
        filtered_results = [
            result for result in similarities
            if result[2] >= settings.similarity_threshold
        ]
        
        return filtered_results[:limit]

File: server/services/vectorService.ts
import { multiAIService } from "./multiAIService";

export async function generateEmbeddings(text: string): Promise<number[]> {
  return await multiAIService.generateEmbeddings(text);
}

export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) {
    throw new Error("Vectors must have the same length");
  }

  let dotProduct = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dotProduct += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

export async function searchSimilarChunks(
  query: string,
  chunks: Array<{ id: string; embedding: number[]; content: string; documentId: string }>,
  topK: number = 5
): Promise<Array<{
  id: string;
  content: string;
  documentId: string;
  similarity: number;
}>> {
  const queryEmbedding = await generateEmbeddings(query);
  
  const similarities = chunks.map(chunk => ({
    ...chunk,
    similarity: cosineSimilarity(queryEmbedding, chunk.embedding),
  }));

  return similarities
    .sort((a, b) => b.similarity - a.similarity)
    .slice(0, topK)
    .map(({ embedding, ...rest }) => rest);
}
